### ISSUE TITLE
unsoundness caused by hidden borrow in nightly match ergonomics

### ISSUE THREAD

**Issue by bvinc** [rust-lang/rust#49631]
[Playground](https://play.rust-lang.org/?gist=49f69f13211329ee026a21fd67a8eb1f&version=nightly)

```rust
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum Foo {
    ReallyReallyReallyLongName,
    Foo2,
    Foo3,
}

struct Iter {
    done: bool
}

impl Iterator for Iter {
    type Item = Result<Foo, String>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.done { return None; }
        self.done = true;
        Some(Ok(Foo::ReallyReallyReallyLongName))
    }
}

fn main() {
    let iter = Iter{done: false};
    let mut iter = iter.peekable();
    
    // peek() returns type Option<&Result<Foo, String>>
    // pattern Some(Ok(foo)) is used instead of Some(&Ok(foo))
    // This causes foo to be a hidden borrow of type &Foo
    while let Some(Ok(foo)) = iter.peek() {
        iter.next();  // destroys what foo is pointing to
        println!("foo={:?}", *foo);
    }
}
```

Output:
```
Illegal instruction (core dumped)
```

The above code should fail the borrow checker.  They key line seems to be the pattern match.  It creates a borrow that seems to be invisible to the borrow checker.  This appears to only be possible in nightly due to the `match_default_bindings` feature.

**Comment by varkor** [comment]
Note that this is fixed with NLL: [playground link](https://play.rust-lang.org/?gist=453f55c9f4fed4f3c136316cf06ecbdb&version=nightly).

**Comment by nikomatsakis** [comment]
Sigh. It works with an explicit ref, as well, which suggests that the problem is the ExprUseVisitor.

**Comment by nikomatsakis** [comment]
OK, so I found the bug. Not sure yet what is best fix. Let me write out what is going on:

- ExprUseVisitor winds up aborting because it invokes the mem-categorization code on a pattern gets back an `Err` result. It assumes (incorrectly) that this indicates a compilation error was reported in typeck.
- mem-categorization is correctly looking at the desugared pattern `Some(&Ok(ref foo))`
  - however, it is *incorrectly* tracking the type of the value being matched
  - that type starts out correct, as `Option<&Result<..>>`
  - but when we pass through the `Some` pattern to the subpatterns:
    - we incorrectly adjust the type to `Result`, not `&Result`
    - this is because we get the type from `pat_ty`:

https://github.com/rust-lang/rust/blob/56714acc5eb0687ed9a7566fdebe5528657fc5b3/src/librustc/middle/mem_categorization.rs#L1265

So we need to stop doing that. Ah, I guess the `pat_adjustments` array stores the information we need:

https://github.com/rust-lang/rust/blob/56714acc5eb0687ed9a7566fdebe5528657fc5b3/src/librustc/ty/context.rs#L368-L381

**Comment by nikomatsakis** [comment]
( Fix in #49714 )
